### PR TITLE
Major refactor (Make things singletons, change some names to be consistent, add simple doc comments to commands, fix formatting issues)

### PR DESCRIPTION
**Pull request by obj-obj**
Commands shouldn't have to have subsystems passed to them

### CODE REVIEW

**Comment by obj-obj** [comment]
Somehow got rid of all my code with the first merge

**Comment by obj-obj** [comment]
This PR can fix that issue because the motors are now set directly through hardware rather than through constructor arguments

**Comment by obj-obj** [comment]
Are the doc comments enough to still communicate which subsystems a command affects?

**Comment by obj-obj** [comment]
They show up pretty cleanly in autocomplete, so I don't think it'll be an issue
Screenshot_20220222_161851

File: src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java [hunk 1]
```diff
@@ -133,6 +105,36 @@ public void setWorkingCommand(boolean working) {
     private double turretTestAngle;
     private double distanceBias;
 
+    /**
+     * Constructor for shooter subsystem.
+     *
+     * @param flywheelMotor1
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java:111]: These parameters got removed, please remove the `@param`s.

File: src/main/java/frc/team2412/robot/commands/shooter/ShooterFlywheelSetVelocityCommand.java [hunk 2]
```diff
@@ -4,8 +4,10 @@
 import frc.team2412.robot.subsystem.ShooterSubsystem;
 
 public class ShooterFlywheelSetVelocityCommand extends InstantCommand {
+    /**
+     * Subsystems: {@link ShooterSubsystem}
+     */
     public ShooterFlywheelSetVelocityCommand(ShooterSubsystem shooter, double velocity) {
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/shooter/ShooterFlywheelSetVelocityCommand.java:10]: Why does this still take a `ShooterSubsystem` parameter?

File: src/main/java/frc/team2412/robot/commands/shooter/ShooterHoodMotorStopCommand.java [hunk 3]
```diff
@@ -4,6 +4,9 @@
 import frc.team2412.robot.subsystem.ShooterSubsystem;
 
 public class ShooterHoodMotorStopCommand extends InstantCommand {
+    /**
+     * Subsystems: {@link ShooterSubsystem}
+     */
     public ShooterHoodMotorStopCommand(ShooterSubsystem shooter) {
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/shooter/ShooterHoodMotorStopCommand.java:10]: Why does this still take a `ShooterSubsystem` parameter?

File: src/main/java/frc/team2412/robot/commands/shooter/ShooterHoodRPMCommand.java [hunk 4]
```diff
@@ -7,6 +7,9 @@
     private final ShooterSubsystem shooter;
     private double rpm, hood;
 
+    /**
+     * Subsystems: {@link ShooterSubsystem}
+     */
     public ShooterHoodRPMCommand(ShooterSubsystem shooter, double targetRPM, double targetHood) {
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/shooter/ShooterHoodRPMCommand.java:13]: Why does this still take a `ShooterSubsystem` parameter?

File: src/main/java/frc/team2412/robot/commands/shooter/ShooterIdleCommand.java [hunk 5]
```diff
@@ -6,6 +6,9 @@
 public class ShooterIdleCommand extends CommandBase {
     public ShooterSubsystem shooterSubsystem;
 
+    /**
+     * Subsystems: {@link ShooterSubsystem}
+     */
     public ShooterIdleCommand(ShooterSubsystem subsystem) {
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/shooter/ShooterIdleCommand.java:12]: Why does this still take a `ShooterSubsystem` parameter?

File: src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java [hunk 6]
```diff
@@ -4,23 +4,21 @@
...
     public enum Bitmap {
         // ingesthasball, feederhasball, ingestcorrectcolor, feedercorrectcolor, intakespeed, ingestspeed,
         // feederspeed, misfire
         // spotless:off
-        A(false, false, INTAKE_IN_SPEED, INDEX_IN_SPEED, INDEX_IN_SPEED, "no balls in system"),
-        B(true, false, INTAKE_IN_SPEED, INDEX_IN_SPEED, INDEX_IN_SPEED, "Ball in Ingest"),
-        C(false, true, INTAKE_IN_SPEED, INDEX_IN_SPEED, 0, "Ball in Feeder"),
-        D(true, true, 0, 0, 0, "Ball in both");
-
+        A(false, false, INTAKE_IN_SPEED, INDEX_IN_SPEED, INDEX_IN_SPEED, "no balls in system"), B(true, false,
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java:19]: Please undo enum formatting here too.

File: src/main/java/frc/team2412/robot/commands/autonomous/SwerveControllerCommand.java [hunk 7]
```diff
@@ -136,25 +113,13 @@ public SwerveControllerCommand(
...
-                trajectory,
-                pose,
-                kinematics,
-                xController,
-                yController,
-                thetaController,
+    public SwerveControllerCommand(Trajectory trajectory, Supplier<Pose2d> pose,
+            SwerveDriveKinematics kinematics, PIDController xController, PIDController yController,
+            ProfiledPIDController thetaController, Consumer<ChassisSpeeds> outputModuleStates,
+            DrivebaseSubsystem drivebase, Subsystem... requirements) {
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/autonomous/SwerveControllerCommand.java:119]: Why does this still have a `DrivebaseSubsystem` parameter?

File: src/main/java/frc/team2412/robot/commands/autonomous/TwoBallAutoCommandLeft.java [hunk 8]
```diff
@@ -16,33 +16,30 @@
...
-            IntakeSubsystem intakeSubsystem) {
-        // Robot should be pressed up on the left side of the lower exit closest to the drivers, facing
+    /**
+     * Subsystems: {@link DrivebaseSubsystem}, {@link IndexSubsystem}, {@link ShooterSubsystem},
+     * {@link ShooterVisionSubsystem}
+     */
+    public TwoBallAutoCommandLeft() {
+        // Robot should be pressed up on the left side of the lower exit closest to the drivers,
+        // facing
         // directly away from the hub with the turret facing towards it
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/autonomous/TwoBallAutoCommandLeft.java:30]: If you want to join wrapped comments, you should join these two lines.

File: src/main/java/frc/team2412/robot/commands/autonomous/TwoBallAutoCommandMiddle.java [hunk 9]
```diff
@@ -16,33 +16,30 @@
...
-            IntakeSubsystem intakeSubsystem) {
-        // Robot should be pressed up on the right side of the lower exit closest to the drivers, facing
+    /**
+     * Subsystems: {@link DrivebaseSubsystem}, {@link IndexSubsystem}, {@link ShooterSubsystem},
+     * {@link ShooterVisionSubsystem}
+     */
+    public TwoBallAutoCommandMiddle() {
+        // Robot should be pressed up on the right side of the lower exit closest to the drivers,
+        // facing
         // directly away from the hub with the turret facing towards it
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/autonomous/TwoBallAutoCommandMiddle.java:30]: If you're joining wrapped comments, you should combine these two lines.

File: src/main/java/frc/team2412/robot/commands/autonomous/TwoBallAutoCommandRight.java [hunk 10]
```diff
@@ -16,33 +16,29 @@
...
-            IntakeSubsystem intakeSubsystem) {
-        // Robot should be pressed up on the left side of the lower exit further from the drivers on their
+    /**
+     * Subsystems: {@link DrivebaseSubsystem}, {@link IndexSubsystem}, {@link ShooterSubsystem},
+     * {@link ShooterVisionSubsystem}
+     */
+    public TwoBallAutoCommandRight() {
+        // Robot should be pressed up on the left side of the lower exit further from the drivers on
+        // their
         // right, facing directly away from the hub with the turret facing towards it
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/autonomous/TwoBallAutoCommandRight.java:30]: If you're joining wrapped comments, you should combine these two lines.

File: src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java [hunk 11]
```diff
@@ -8,29 +8,25 @@
...
 public class IntakeBitmapCommand extends CommandBase {
-
     public final double MISFIRE_VELOCITY = 400;
 
     // bitmap
     public enum Bitmap {
-        // ingesthasball, feederhasball, ingestcorrectcolor, feedercorrectcolor, intakespeed, ingestspeed,
+        // ingesthasball, feederhasball, ingestcorrectcolor, feedercorrectcolor, intakespeed,
+        // ingestspeed,
         // feederspeed, misfire
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/commands/intake/IntakeBitmapCommand.java:17]: If you want to join wrapped comments, you should combine these two lines.

File: src/main/java/frc/team2412/robot/sim/SparkMaxSimProfile.java [hunk 12]
```diff
@@ -18,9 +18,11 @@
...
     }
 
-    // Note: REVPhysicsSim is supposed to have a singleton, but instantiating separate instances works.
+    // Note: REVPhysicsSim is supposed to have a singleton, but instantiating separate instances
+    // works.
     // Each sim has a separate ArrayList, though, so having separate sims with only one motor is
-    // inefficient. The reason to have separate instances for each sim profile is so that running one
+    // inefficient. The reason to have separate instances for each sim profile is so that running
+    // one
     // won't affect others.
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/sim/SparkMaxSimProfile.java:26]: If you want to join wrapped comments, you should combine these lines.

File: src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java [hunk 13]
```diff
@@ -166,10 +156,10 @@ private void configMotors() {
         hoodMotor.setInverted(true);
         hoodMotor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kForward, true);
         hoodMotor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, true);
+        // Current hood setup plan start hood at 0, below MIN_HOOD_ANGLE
         hoodMotor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward,
                 (float) (MAX_HOOD_ANGLE / HOOD_REVS_TO_DEGREES));
-        hoodMotor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, 0); // Current hood setup plan starts hood at 0,
-                                                                            // below MIN_HOOD_ANGLE
```
Comment (KangarooKoala) [inline src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java:172]: The comment is specific to setting the reverse soft limit.

File: src/main/java/frc/team2412/robot/commands/autonomous/TwoBallAutoCommandLeft.java (hunk 8 above)
Comment (obj-obj) [reply src/main/java/frc/team2412/robot/commands/autonomous/TwoBallAutoCommandLeft.java:30]: Strange, I set formatter to automatically join lines

File: src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java (hunk 13 above)
Comment (obj-obj) [reply src/main/java/frc/team2412/robot/subsystem/ShooterSubsystem.java:172]: How did i move that comment to the wrong line...

**Comment by obj-obj** [comment]
Ok I think this is ready now, I just checked through all the commands and I don't think I've missed anything else

I don't think removing the subsystem arguments from command is an issue, because doc comments show up just fine on vscode's autocomplete/hover, allowing you to still easily see what subsystems a command uses. Also, isn't this what documentation is for..?
Also, it's just annoying to have to pass in multiple subsystems when running some commands (it's just extra boilerplate).
As for reviewing PRs, just install the GitHub extension in vscode, then go to the github tab on the sidebar and click one of the PRs to open it in vscode.

**Review by bobtseattle — CHANGES_REQUESTED** [review]
There is too much going in this PR.  
Please split it up into smaller PRs to make it easier to review, merge, and / or rollback.
There are some good things in this PR, but others require more thought, discussion and buy-in from the team.  We shouldn't block the good things while waiting to reach consensus on the others.  Breaking the PR up into more manageable, target PRs will be helpful for everyone.

**Comment by KangarooKoala** [comment]
I think (correct me if I'm wrong) there are three major changes:

1. Singletons

    Changed `Controls`, `Hardware`, `Robot` (changed implementation), `Subsystems`, and the subsystems to singletons
    Added doc comments to commands about which subsystems they used

2. Formatting

    Fixed enum formatting, change line length to 110, insert newlines at ends of files, and join lines

3. Names

    Renamed `climbMotorFixed` to `climbFixedMotor`, `climbMotorDynamic` to `climbDynamicMotor`, `ingestIndexMotor` to `ingestMotor`, `feederIndexMotor` to `indexFeederMotor`

**Comment by obj-obj** [comment]
Yeah I just kind of threw things in here

**Comment by obj-obj** [comment]
I can probably use git cherrypick to split everything apart without rewriting

**Pull request closed without merging**